s3_sync/utils.py

```
@@ -73,61 +73,174 @@ def __unicode__(self):
         return u'Error (%d): %s' % (self.resp.status, self.resp.body)
 
 
-class FileWrapper(object):
+class SeekableFileLikeIter(FileLikeIter):
+    """
+    Like Swift's FileLikeIter, with the following changes in/additions of
+    behavior:
+        * You can optionally specify a length, and reads past that length will
+          return data only up to that length, and EOF after that.
+        * If a length is specified, len() will work on this object, otherwise
+          it will raise a UnsupportedOperation.
+        * If you specify a `seek_zero_cb`, it will be called when a seek to
+          position zero (regardless of args), if any data has already been
+          read.  The callback must return a new iterator, which will be
+          used for subsequent reads.  In other words, the callback must
+          actually take some action that allows the data coming out to be
+          correct for an actual offset of zero.
+    """
+    def __init__(self, iterable, length=None, seek_zero_cb=None):
+        super(SeekableFileLikeIter, self).__init__(iterable)
+        self.length = length
+        self.seek_zero_cb = seek_zero_cb
+        self._bytes_delivered = 0  # capped by length, if given
+
+    def tell(self):
+        return self._bytes_delivered
+
+    def _length_exceeded(self):
+        if self.length is None:
+            return False
+        bytes_we_can_return = self.length - self.tell()
+        return not bytes_we_can_return
+
+    def seek(self, pos, flag=0):
+        if pos != 0:
+            raise RuntimeError('%s: arbitrary seeks are not supported' %
+                               self.__class__.__name__)
+        if self._bytes_delivered == 0:
+            return
+        if self.seek_zero_cb:
+            self.iterator = self.seek_zero_cb()
+            self.buf = None
+        # Without a callback, we leave self.buf alone (there's no truly "right"
```

Is there a use case where we want to allow resetting the stream without a callback? Maybe we should raise an exception instead, as otherwise the returned data may actually be corrupted and not at all what the caller expects.

---

will fix